File: crates/workspace-core/src/restore.rs
use crate::SessionTransport;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RestoreRecipe {
    pub launch_command: String,
}

pub fn build_restore_recipe(
    transport: SessionTransport,
    target_label: &str,
    shell: &str,
    cwd: Option<&str>,
) -> RestoreRecipe {
    let shell_command = match cwd {
        Some(cwd) => format!("cd {} && exec {} -l", shell_quote(cwd), shell_quote(shell)),
        None => format!("exec {} -l", shell_quote(shell)),
    };

    let launch_command = match transport {
        SessionTransport::Local => shell_command,
        SessionTransport::Ssh => format!(
            "ssh {target_label} -- '{}'",
            escape_for_single_quotes(&shell_command)
        ),
    };

    RestoreRecipe { launch_command }
}

fn shell_quote(value: &str) -> String {
    if value
        .chars()
        .all(|ch| ch.is_ascii_alphanumeric() || matches!(ch, '/' | '.' | '_' | '-' | ':'))
    {
        value.to_owned()
    } else {
        format!("'{}'", escape_for_single_quotes(value))
    }
}

fn escape_for_single_quotes(value: &str) -> String {
    value.replace('\'', "'\"'\"'")
}

File: libs/workspace-core/src/restore.zig
const std = @import("std");
const models = @import("models.zig");

pub fn buildRestoreRecipe(
    allocator: std.mem.Allocator,
    transport: models.SessionTransport,
    target_label: []const u8,
    shell: []const u8,
    cwd: ?[]const u8,
) !models.RestoreRecipe {
    const quoted_shell = try shellQuote(allocator, shell);
    defer allocator.free(quoted_shell);

    const shell_command = if (cwd) |value| blk: {
        const quoted_cwd = try shellQuote(allocator, value);
        defer allocator.free(quoted_cwd);
        break :blk try std.fmt.allocPrint(allocator, "cd {s} && exec {s} -l", .{
            quoted_cwd,
            quoted_shell,
        });
    } else try std.fmt.allocPrint(allocator, "exec {s} -l", .{
        quoted_shell,
    });
    defer allocator.free(shell_command);

    const launch_command = switch (transport) {
        .local => try allocator.dupe(u8, shell_command),
        .ssh => blk: {
            const escaped = try escapeForSingleQuotes(allocator, shell_command);
            defer allocator.free(escaped);
            break :blk try std.fmt.allocPrint(allocator, "ssh {s} -- '{s}'", .{
                target_label,
                escaped,
            });
        },
    };

    return .{ .launch_command = launch_command };
}

fn shellQuote(allocator: std.mem.Allocator, value: []const u8) ![]u8 {
    for (value) |ch| {
        if (!std.ascii.isAlphanumeric(ch) and ch != '/' and ch != '.' and ch != '_' and ch != '-' and ch != ':') {
            const escaped = try escapeForSingleQuotes(allocator, value);
            defer allocator.free(escaped);
            return std.fmt.allocPrint(allocator, "'{s}'", .{escaped});
        }
    }
    return allocator.dupe(u8, value);
}

fn escapeForSingleQuotes(allocator: std.mem.Allocator, value: []const u8) ![]u8 {
    var buffer: std.ArrayList(u8) = .empty;
    defer buffer.deinit(allocator);

    for (value) |ch| {
        if (ch == '\'') {
            try buffer.appendSlice(allocator, "'\"'\"'");
        } else {
            try buffer.append(allocator, ch);
        }
    }

    return buffer.toOwnedSlice(allocator);
}
